Update the handlePush method for $opened_push event

evan-masseau opened the pull request:
Pertains to [TP148025](https://klaviyo.tpondemand.com/entity/148025-update-opened-push-event) so that push token can be added to the Opened Push event's properties.

Updated Opened Push to (intentionally) send whole push payload back as properties. Serendipitously we already were doing that because the previous code was checking for "_k" in the wrong place anyways.

Added push token to the Opened Push track event payload directly (is it going to be removed from customer properties?) 

And added a suggestion for how to manage access to the push token as a private property with getter/setter function that handle persistence to/from UserDefaults

ndurell reviewed (commented):
This would work however we will most likely end up storing this a different way.

evan-masseau commented:
OK, at the end of the day adding push token to that payload is all I really needed. You did say the push token is going to be removed from customer properties, right?

ajaysubra commented on Sources/KlaviyoSwift/Klaviyo.swift:
```diff
@@ -167,16 +168,18 @@ public class Klaviyo : NSObject {
-        } else {
-            trackEvent(eventName: KLPersonOpenedPush, properties: userInfo as NSDictionary)
+        // Test it originated from klaviyo by checking presence of _k tracking parameters
+        if let body = userInfo["body"] as? NSDictionary, let _ = body["_k"] as? NSDictionary {
```
❔ Curious why do we need to type cast  `body["_k"]` to a `NSDictionary` in order to see if body has the key `_k`?

evan-masseau replied:
I suppose we don't, other than to make sure _k is a dict as we'd expect. When we circle back to this after noah's bigger changes, I can revisit

ajaysubra commented on Sources/KlaviyoSwift/Klaviyo.swift:
```diff
@@ -167,16 +168,18 @@ public class Klaviyo : NSObject {
+        // Test it originated from klaviyo by checking presence of _k tracking parameters
+        if let body = userInfo["body"] as? NSDictionary, let _ = body["_k"] as? NSDictionary {
+            //But we still want to send the full payload, plus the push token
+            let properties = userInfo.mutableCopy() as! NSMutableDictionary
```
❔ Are there any risk associated with force type casting `userInfo.mutableCopy()` to a `NSMutableDictionary` like a crash? Can we conditionally  type cast it?

evan-masseau replied:
I don't think it should ever fail, but fair enough, it could be another condition in the if let above.

ajaysubra commented on Sources/KlaviyoSwift/Klaviyo.swift:
```diff
@@ -331,17 +334,31 @@ public class Klaviyo : NSObject {
+     - Returns: String?
+     */
+    public func getPushDeviceToken() -> String? {
+        if apnDeviceToken == nil {
```
💭  Does it make sense to keep `UserDefaults` the source of truth for getting the device token instead of the in memory variable just so that we don't have to share the variable between functions here?

ajaysubra replied:
If we did this this function could be a simple read only computed property. Just a suggestion.

evan-masseau replied:
Just a little quicker to have it in memory too is all

ajaysubra reviewed (commented):
Added some minor questions but nothing that should block you.

ndurell commented:
> OK, at the end of the day adding push token to that payload is all I really needed. You did say the push token is going to be removed from customer properties, right?

Yah I don't think it makes sense to include in every request how I will ensure it gets included in the open event as a part of the payload.

evan-masseau commented:
Updated to remove conflict after Noah's big PRs were merged... but most of what I had changed is just totally moved around now so the only purpose of the PR at this point is just to make sure push_token gets added to the $opened_push track event

evan-masseau commented on Sources/KlaviyoSwift/Klaviyo.swift:
```diff
@@ -139,14 +139,20 @@ public class Klaviyo: NSObject  {
+        // Test it originated from klaviyo by checking presence of _k tracking parameters
+        if let body = userInfo["body"] as? NSDictionary, let _ = body["_k"] as? NSDictionary {
+            //But we still want to send the full payload, plus the push token
+            _ = environment.analytics.store.state.sink{ completion in
```
I need to read up on this composed architecture pattern some more, maybe this isn't the proper way to access current state?

evan-masseau commented:
Unneeded.
